### ISSUE TITLE
Single PWA with i18n

### ISSUE THREAD

**Issue by TheSlimvReal** [angular/angular#43796]
### Which @angular/* package(s) are relevant/releated to the feature request?

localize, service-worker

### Description

Following up on [this post](https://stackoverflow.com/q/61532813/10713841) on stackoverflow. 

After adding internationalization (`i18n`) to our app we run into the big problem of backwards compatibility. Before adding `i18n` we had our service worker at the root folder and after adding `i18n` we now have a separate service worker for each language we translated the app into. This is a problem because a user that installed the app prior to this change will not correctly get updates because the old service worker will not be able to find the new service worker.

This is the default behavior of Angular and left us with the decision that we either have to tell all our users that they will have to uninstall and re-install the app or add a service worker in the root folder that automatically takes care of the migration and leave it deployed until all users are migrated. Theses solutions are rather unsatisfying, poorly documented and mean that a we would have to go through the same struggle again if we decide to remove the `i18n` or take a different approach for it.

### Proposed solution

It would be great if Angular would not enforce to use the _one-PWA-per-language_ approach but rather leave this as a choice to the developer. This would make the transition from a single- to a multi-language app much easier and leaves the option open to also go back to the previous approach (no `i18n`).

This could be configured through the compiler options in the `angular.json` file. E.g. `rootServiceWorker: true`.   

### Alternatives considered

Alternatively, it would be great if Angular provided a guide or a predefined service worker (similar to the `safety-worker.js`) that allows a easy migration of the PWAs when `i18n` is introduced.

**Comment by petebacondarwin** [comment]
Hi @TheSlimvReal - sorry not to respond so far on this issue. It is a bit tricky to implement on our side. I need to talk with @gkalpak about how this could work - I am not sure if it is something we would do. I am glad that for the time being you appear to have a workaround.

**Comment by Kamshak** [comment]
If you load your translation files at runtime using @angular/localize you can have a single pwa with multiple languages. You can still lazy load the translation files so that all languages are not part of your main bundle

**Comment by TheSlimvReal** [comment]
> If you load your translation files at runtime using @angular/localize you can have a single pwa with multiple languages. You can still lazy load the translation files so that all languages are not part of your main bundle

@Kamshak What exactly are you revering to when you talk about _at runtime_? `@angular/localize` uses the translation files at build time and creates a single application for each language. The translation files are not part of the final build.

**Comment by Kamshak** [comment]
> > If you load your translation files at runtime using @angular/localize you can have a single pwa with multiple languages. You can still lazy load the translation files so that all languages are not part of your main bundle
> 
> @Kamshak What exactly are you revering to when you talk about _at runtime_? `@angular/localize` uses the translation files at build time and creates a single application for each language. The translation files are not part of the final main bundle.

Yes, you can build separate application bundles for each language (inlining the translations). You use `@angular/localize` only for tagging strings. As a postbuild step the CLI will basically get rid of all the `$localize` calls in your code by replacing the calls to it with the translations more or less directly. The issue is you need to now serve separate PWAs for each language because you essentially have different apps with different file hashes and everything.

Another option is to ship translation files with your app and load them at run time (i.e. generate one app bundle +  multiple files containing only the messages). That way Angular will keep the actual calls to `$localize` in your app, which then perform the translations when the app is running. To do this you include `@angular/localize/init` and call `loadTranslations` at runtime. Now what I'm referring to is this: you need to have messages available so you want to ship the translation file with your app. But you don't want to include ALL languages into your main bundle. So you can write a small bit of bootstrap logic to get the user's language, then load only the needed translation file on demand before the app bootstraps. I can do a small writeup on how to wire this up exactly if there's interest

**Comment by TheSlimvReal** [comment]
@Kamshak Thank you for pointing that out, I had no idea that this `loadTranslations` function exists. I guess I have to try out whether the performance is still alright. At least it keeps the PWA size smaller. Have you ever worked with that approach?

**Comment by Kamshak** [comment]
> @Kamshak Thank you for pointing that out, I had no idea that this `loadTranslations` function exists. I guess I have to try out whether the performance is still alright. At least it keeps the PWA size smaller. Have you ever worked with that approach?

Have a PWA in production with this, it's working great so far

**Comment by TheSlimvReal** [comment]
@Kamshak Sorry for the long silence. My previously mentioned workaround was working quite well but now we are running into some new problems regarding the service-workers and I would like to try out your approach.

Just out of curiosity: How do you create your translation files for the `loadTranslations` function? Do you use the `xlf` format and then parse it using something like the merge script described [here](https://developapa.com/angular-localize/#setup)?
It would be great if you could point me to a reference implementation where `loadTranslations` is used as the docs are very sparse on that point (especially regarding caching and bootstrap).

**Comment by perkoe** [comment]
Hi @TheSlimvReal and @ValentinFunk,

I've encountered similar issues with localization and service worker integration, particularly around Angular's built-in internationalization and PWA caching strategies.

You mentioned previously running into additional problems with your original workaround related to service workers, and then moving toward Angular's loadTranslations method instead. Could you elaborate briefly on:

1. What specific new issues did you face regarding service workers with the initial approach (the merge-script method)?

2. Did moving to the loadTranslations method completely solve those issues for you, especially regarding caching, bootstrap performance, and size?

3. Could you share a reference implementation or briefly outline how you structured your translation files, managed caching, and handled bootstrap performance with loadTranslations?

I'd greatly appreciate any insights or examples you can provide, as official documentation is sparse and your experience would be incredibly valuable.

Additionally, I noticed your original merge script **doesn't copy the index.htm** from the primary locale into the root directory (dist/webapp/browser). Was there a specific reason behind leaving index.html localized rather than placing a default one at the root?

I'm curious if this was intentional for handling locales separately or if it was an oversight. How did you manage the root-level index.html serving when using this approach?

**Comment by JeanMeche** [comment]
With #63364 we've announced that the service-worker package in now in maintenance mode and will not receive feature PRs and feature requests.

The message is also visible here: https://angular.dev/ecosystem/service-workers
